Common/ImageSamplers/itkImageSamplerBase.hxx

```diff
@@ -443,6 +444,247 @@ ImageSamplerBase<TInputImage>::PrintSelf(std::ostream & os, Indent indent) const
+  splitRegion.SetSize(splitSize);
+
+  return maxThreadIdUsed + 1;
+
+} // end SplitRegion()
+
+
+/**
+ * ******************* GenerateData *******************
+ */
+
+template <class TInputImage>
+void
+ImageSamplerBase<TInputImage>::GenerateData()
+{
+  // Call a method that can be overriden by a subclass to allocate
+  // memory for the filter's outputs
+  // this->AllocateOutputs();
+
+  // Call a method that can be overridden by a subclass to perform
+  // some calculations prior to splitting the main computations into
+  // separate threads
+  this->BeforeThreadedGenerateData();
+
+  // Set up the multithreaded processing
+  ThreadStruct str;
+  str.Filter = this;
+
+  this->GetMultiThreader()->SetNumberOfWorkUnits(this->GetNumberOfWorkUnits());
+  this->GetMultiThreader()->SetSingleMethod(this->ThreaderCallback, &str);
+
+  // multithread the execution
+  this->GetMultiThreader()->SingleMethodExecute();
+
+  // Call a method that can be overridden by a subclass to perform
+  // some calculations after all the threads have completed
+  this->AfterThreadedGenerateData();
+
+} // end GenerateData()
+
+
+/**
+ * ******************* ThreadedGenerateData *******************
+ */
+
+template <class TInputImage>
+void
+ImageSamplerBase<TInputImage>::ThreadedGenerateData(const InputImageRegionType &, ThreadIdType)
+{
+  // The following code is equivalent to:
+  // itkExceptionMacro("subclass should override this method!!!");
+  // The ExceptionMacro is not used because gcc warns that a
+  // 'noreturn' function does return
+  std::ostringstream message;
+  message << "itk::ERROR: " << this->GetNameOfClass() << "(" << this << "): Subclass should override this method!!!";
+  ExceptionObject e_(__FILE__, __LINE__, message.str().c_str(), ITK_LOCATION);
+  throw e_;
+
+} // end ThreadedGenerateData()
+
+
+/**
+ * ******************* ThreaderCallback *******************
+ */
+
+// Callback routine used by the threading library. This routine just calls
+// the ThreadedGenerateData method after setting the correct region for this
+// thread.
+template <class TInputImage>
+ITK_THREAD_RETURN_FUNCTION_CALL_CONVENTION
+ImageSamplerBase<TInputImage>::ThreaderCallback(void * arg)
+{
+  assert(arg);
+  const auto &       workUnitInfo = *static_cast<const MultiThreaderBase::WorkUnitInfo *>(arg);
+  const ThreadIdType threadId = workUnitInfo.WorkUnitID;
+  const ThreadIdType threadCount = workUnitInfo.NumberOfWorkUnits;
+
+  assert(workUnitInfo.UserData);
+  const auto & str = *static_cast<const ThreadStruct *>(workUnitInfo.UserData);
+
+  assert(str.Filter);
+  Self & filter = *(str.Filter);
+
+  // execute the actual method with appropriate output region
+  // first find out how many pieces extent can be split into.
+  // \todo: requested region -> this->GetCroppedInputImageRegion()
+  typename TInputImage::RegionType splitRegion;
+  const unsigned int total = SplitRegion(filter.GetInput()->GetRequestedRegion(), threadId, threadCount, splitRegion);
```

Comment:
FYI, here in  in `ThreaderCallback` is where I would like to be able to use `ImageSamplerBase::m_CroppedInputImageRegion`, instead of the requested region.